### PR TITLE
refactor(starknet_batcher): add tests and refactor batcher_test

### PR DESCRIPTION
**Pull request by dafnamatsry**

### CODE REVIEW

**Comment by dafnamatsry** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2640?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2640** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2640?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2640?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#2618** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2618?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2622** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2622?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2441** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2441?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2483** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2483?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2480** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2480?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by Yael-Starkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2640)* status: 0 of 3 files reviewed, 9 unresolved discussions (waiting on @alonh5 and @dafnamatsry)

___
*[`crates/starknet_batcher/src/batcher_test.rs` line 242 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OESPIIiC7H-V9SsnVMi:-OESPIIiC7H-V9SsnVMj:b-td1v3f) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L242)):*
> ```XML
> #[rstest]
> #[case::send_txs(SendProposalContent::Txs(test_txs(0..1)))]
> #[case::send_finish(SendProposalContent::Finish)]
> ```

add a case for SendProposalContent::Abort



_Code quote:_
```XML
#[case::send_txs(SendProposalContent::Txs(test_txs(0..1)))]
#[case::send_finish(SendProposalContent::Finish)]
```
___
*[`crates/starknet_batcher/src/batcher_test.rs` line 254 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OESPV6R26lMrcDPPp4h:-OESPV6R26lMrcDPPp4i:bnfhta7) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L254)):*
> ```XML
> #[rstest]
> #[case::send_txs(SendProposalContent::Txs(test_txs(0..1)))]
> #[case::send_finish(SendProposalContent::Finish)]
> ```

what happens if you abort an invalid proposal?

File: crates/starknet_batcher/src/batcher_test.rs [hunk 1]
```diff
@@ -229,106 +206,140 @@ async fn no_active_height() {
...
+    let first_propose_block_input = propose_block_input(PROPOSAL_ID);
+    let mut second_propose_block_input = first_propose_block_input.clone();
+    second_propose_block_input.block_info.block_number = INITIAL_HEIGHT.unchecked_next();
 
+    // Start the first height and propose block.
     batcher.start_height(StartHeightInput { height: INITIAL_HEIGHT }).await.unwrap();
-    batcher.validate_block(validate_block_input()).await.unwrap();
+    batcher.propose_block(first_propose_block_input).await.unwrap();
+
+    // Start the second height, and make sure the previouse height proposal is cleared, by trying to
```
[the remark this answers is not in the document]
Comment (dafnamatsry) [reply crates/starknet_batcher/src/batcher_test.rs]: Done.

**Review by dafnamatsry — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2640)* status: 0 of 3 files reviewed, 10 unresolved discussions (waiting on @alonh5, @graphite-app[bot], and @Yael-Starkware)

___
*[`crates/starknet_batcher/src/batcher_test.rs` line 123 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OEOfRFZBauGRv6d6YJv:-OEhSR5O1MGEhikfScut:b3sutxf) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L123)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

why do we need the new object? Is any other module using it?
</blockquote></details>

It's dedicated to the `batcher_test`.
It replaces the need to spawn an async task that simulates the block builder, with a fake block builder that does the same.
It's not a must, but IMO just nicer.
___
*[`crates/starknet_batcher/src/batcher_test.rs` line 167 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OEOmUlt0nEfRbN3sJV2:-OEhbZUQ1cFKQhx6ppBN:bqmjwcz) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L167)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

Is there a test that checks that after start_height , the previous height proposal are not longer present?
</blockquote></details>

Added.
___
*[`crates/starknet_batcher/src/batcher_test.rs` line 242 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OESPIIiC7H-V9SsnVMi:-OEhV3qvCX0Vo8Vamzzj:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L242)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

add a case for SendProposalContent::Abort
</blockquote></details>

Done.
___
*[`crates/starknet_batcher/src/batcher_test.rs` line 254 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OESPV6R26lMrcDPPp4h:-OEhWWJgA4GQWCcLrsTU:b-5u3yav) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L254)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

what happens if you abort an invalid proposal?
</blockquote></details>

Added.
We should get `Aborted`.
___
*[`crates/starknet_batcher/src/batcher_test.rs` line 267 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OEOi1ps9Tkdt5y-MYhA:-OEhYKwkEwBsB1Hv_2pd:bitjhth) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L267)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

we should also try to send txs and finish to the aborted proposal , should they get Status::aborted or an error in return?

also, what would a second abort request get? an error or another status aborted?
</blockquote></details>

We should get a `ProposalNotFound` error, for any `send_proposal_content` request we get after we got `Finish` or `Abort`. We shouldn't get requests like that from the consensus.
___
*[`crates/starknet_batcher/src/batcher_test.rs` line 390 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OEOit5E7n1VhqqLDm7h:-OEhapk-6BUBqX2xdvZ1:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L390)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

originally in the proposal manager we had 4 interleaving calls - propose, validate, propose, validate, to make sure that propose works after validate, validate after propose , propose after propose and validate after validate .
</blockquote></details>

Done.
___
*[`crates/starknet_batcher/src/batcher_test.rs` line 402 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OEOjOan7NhWCNTm5PG3:-OEharX_CHHiVzO2mbXp:byzugmh) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L402)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

I don't think this error best describes the scenario.
The server is not really busy, it is a consensus bug that it started a validate_block flow and didn't finish it.
</blockquote></details>

Will be changed in a separate PR.
___
*[`crates/starknet_batcher_types/src/batcher_types.rs` line 34 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OEOjztZ6ULknjbZ2XrS:-OEhbXRpAJfAS4p6JQ3Z:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher_types/src/batcher_types.rs#L34)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

where is this default being used?
</blockquote></details>

Done.
___
*[`crates/starknet_batcher_types/src/batcher_types.rs` line 64 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OEOk2jk9gob6jL1_2Fu:-OEhbXusDhRMIx_E6ZYi:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher_types/src/batcher_types.rs#L64)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

same.
</blockquote></details>

Done.

File: crates/starknet_batcher/src/batcher_test.rs [hunk 2]
```diff
@@ -229,106 +206,140 @@ async fn no_active_height() {
...
+        );
+    }
+
+    let mut batcher = create_batcher(MockDependencies {
+        block_builder_factory,
+        storage_reader,
+        ..Default::default()
+    });
+
+    // Prepate the propose_block requests for the first and the second heights.
```
[the remark this answers is not in the document]
Comment (dafnamatsry) [reply crates/starknet_batcher/src/batcher_test.rs]: Done.

File: crates/starknet_batcher/src/batcher_test.rs [hunk 3]
```diff
@@ -399,6 +410,53 @@ async fn get_content_from_unknown_proposal() {
...
+            &mut block_builder_factory,
+            Ok(BlockExecutionArtifacts::create_for_testing()),
+        );
+    }
+    let mut batcher =
+        create_batcher(MockDependencies { block_builder_factory, ..Default::default() });
+
+    batcher.start_height(StartHeightInput { height: INITIAL_HEIGHT }).await.unwrap();
+
+    // Make sure we can generate 4 consequtive proposals.
```
[the remark this answers is not in the document]
Comment (dafnamatsry) [reply crates/starknet_batcher/src/batcher_test.rs]: Done.

**Review by dafnamatsry — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2640)* status: 0 of 3 files reviewed, 12 unresolved discussions (waiting on @alonh5, @graphite-app[bot], and @Yael-Starkware)

File: crates/starknet_batcher/src/batcher_test.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Yael-Starkware) [reply crates/starknet_batcher/src/batcher_test.rs]: cool!

**Review by Yael-Starkware — CHANGES_REQUESTED** [review]
> Reviewed 1 of 3 files at r1, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2640)* status: 2 of 3 files reviewed, 7 unresolved discussions (waiting on @alonh5, @dafnamatsry, and @graphite-app[bot])

___
*[`crates/starknet_batcher/src/batcher_test.rs` line 123 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OEOfRFZBauGRv6d6YJv:-OEiXBSu5eZAPChRIOBv:b-ajkcfa) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L123)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

It's dedicated to the `batcher_test`.
It replaces the need to spawn an async task that simulates the block builder, with a fake block builder that does the same.
It's not a must, but IMO just nicer.
</blockquote></details>

ok, I prefer the short version, but it's a matter of preference , so choose to our preference.
___
*[`crates/starknet_batcher/src/batcher_test.rs` line 267 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OEOi1ps9Tkdt5y-MYhA:-OEiZTjA9GR9YTn_oGRR:by02a2i) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L267)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

We should get a `ProposalNotFound` error, for any `send_proposal_content` request we get after we got `Finish` or `Abort`. We shouldn't get requests like that from the consensus.
</blockquote></details>

is it covered in the tests?
___
*[`crates/starknet_batcher/src/batcher_test.rs` line 402 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OEOjOan7NhWCNTm5PG3:-OEiY99N6iljBEWyerdO:b57a9xh) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L402)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Will be changed in a separate PR.
</blockquote></details>

can you add a todo?

File: crates/starknet_batcher/src/batcher_test.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dafnamatsry) [reply crates/starknet_batcher/src/batcher_test.rs]: Done.

**Review by dafnamatsry — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2640)* status: 1 of 4 files reviewed, 6 unresolved discussions (waiting on @alonh5, @graphite-app[bot], and @Yael-Starkware)

___
*[`crates/starknet_batcher/src/batcher_test.rs` line 267 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OEOi1ps9Tkdt5y-MYhA:-OEmIMBCAwLTnb-oeUhw:bwhu7y) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L267)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

is it covered in the tests?
</blockquote></details>

Now it is.
___
*[`crates/starknet_batcher/src/batcher_test.rs` line 402 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-OEOjOan7NhWCNTm5PG3:-OEmKyZ-67gR0cbYKgBj:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/63fc881aa4777c382f456fecb843a9076ee69563/crates/starknet_batcher/src/batcher_test.rs#L402)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

can you add a todo?
</blockquote></details>

Done.

**Review by Yael-Starkware — COMMENTED** [review]
> Reviewed 1 of 2 files at r8.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2640)* status: 2 of 4 files reviewed, 3 unresolved discussions (waiting on @alonh5 and @graphite-app[bot])

**Review by Yael-Starkware — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2640#-:-OErWXYI1e8cdGyqYDal:bnfp4nl)* status: 2 of 4 files reviewed, 3 unresolved discussions (waiting on @alonh5 and @graphite-app[bot])

**Review by Yael-Starkware — APPROVED** [review]
> Reviewed 1 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2640)* status: 3 of 4 files reviewed, 3 unresolved discussions (waiting on @alonh5 and @graphite-app[bot])

**Review by dafnamatsry — COMMENTED** [review]
> Dismissed @graphite-app[bot] from 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2640)* status: 3 of 4 files reviewed, all discussions resolved (waiting on @alonh5)

**Review by dafnamatsry — COMMENTED** [review]
> Reviewed 1 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2640)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @alonh5)

**Pull request merged**